src/tests/conformance/handshake.rs

```diff
@@ -32,3 +36,38 @@ async fn c001_handshake_when_node_receives_connection() {
     synthetic_node.shut_down().await;
     node.stop().unwrap();
 }
+
+#[tokio::test]
+async fn c002_handshake_when_node_initiates_connection() {
+    // ZG-CONFORMANCE-002
+
+    // Create a synthetic node and enable handshaking.
+    let synthetic_node = SyntheticNodeBuilder::default()
+        .build()
+        .await
+        .expect("Unable to build a synthetic node");
+
+    // Spin up a node instance.
+    let target = TempDir::new().expect("Couldn't create a temporary directory");
+    let mut node = Node::builder()
+        .initial_peers(vec![synthetic_node.listening_addr().unwrap()])
+        .build(target.path())
+        .expect("Unable to build the node");
+    node.start().await;
+
+    let node_addr = timeout(CONNECTION_TIMEOUT, synthetic_node.wait_for_connection())
+        .await
+        .expect("Couldn't establish a connection");
+
+    // Check the connection has been established (this is only set post-handshake). We can't check
+    // for the addr as nodes use ephemeral addresses when initiating connections.
+    assert_ne!(node_addr, node.net_addr().unwrap());
+
+    // The node sends multiple get_block HTTP queries from different TCP sockets in parallel,
+    // so in rare ocassions we might have additional few short lasting connections.
+    assert!(synthetic_node.num_connected() >= 1);
```

Comment:
This is not typical behaviour. I'm gonna try to document this in Notes.md.

Here's how it looks like in logs:



```
2022-10-17T10:42:12.955611Z  INFO node{name="0"}: ziggurat_algorand::protocol::handshake: 127.0.0.1:54980: received handshake message: b"GET /v1/private-v1/gossip HTTP/1.1\r\nHost: 127.0.0.1:54978\r\nUser-Agent: algod/3.9 (stable; commit=921e8f6f; 4) darwin(arm64)\r\nConn
ection: Upgrade\r\nSec-WebSocket-Key: r/lmQVxIFwzYtSNLnm9Lug==\r\nSec-WebSocket-Version: 13\r\nUpgrade: websocket\r\nX-Algorand-Accept-Version: 2.1\r\nX-Algorand-Instancename: nASdpsibljD4Fq7Z\r\nX-Algorand-Location: http://127.0.0.1:54979\r\nX-Algorand-Noderandom: D2zA9G
J0QF0RtA==\r\nX-Algorand-Telid: \r\nX-Algorand-Version: 2.1\r\n\r\n"
{"Context":"sync","file":"universalFetcher.go","function":"github.com/algorand/go-algorand/catchup.(*HTTPFetcher).getBlockBytes","level":"debug","line":217,"msg":"block GET \"http://127.0.0.1:54978/v1/private-v1/block/3\" peer \u0026network.wsPeerCore{net:(*network.Websoc
ketNetwork)(0x1400050fc00), rootURL:\"127.0.0.1:54978\", originAddress:\"\", client:http.Client{Transport:(*network.rateLimitingTransport)(0x14000510250), CheckRedirect:(func(*http.Request, []*http.Request) error)(nil), Jar:http.CookieJar(nil), Timeout:0}} *network.wsPeer
Core","name":"127.0.0.1:0","time":"2022-10-17T12:42:12.955347+02:00"}
{"Context":"sync","file":"universalFetcher.go","function":"github.com/algorand/go-algorand/catchup.(*HTTPFetcher).getBlockBytes","level":"debug","line":217,"msg":"block GET \"http://127.0.0.1:54978/v1/private-v1/block/f\" peer \u0026network.wsPeerCore{net:(*network.Websoc
ketNetwork)(0x1400050fc00), rootURL:\"127.0.0.1:54978\", originAddress:\"\", client:http.Client{Transport:(*network.rateLimitingTransport)(0x14000510250), CheckRedirect:(func(*http.Request, []*http.Request) error)(nil), Jar:http.CookieJar(nil), Timeout:0}} *network.wsPeer
Core","name":"127.0.0.1:0","time":"2022-10-17T12:42:12.955373+02:00"}
2022-10-17T10:42:12.955913Z  INFO node{name="0"}: ziggurat_algorand::protocol::handshake: sending handshake message: b"HTTP/1.1 101 Switching Protocols\r\nUpgrade: websocket\r\nConnection: Upgrade\r\nSec-Websocket-Accept: gf5FrYS05FHPScWZ9lVxviOXmuc=\r\nX-Algorand-Instanc
ename: synth_node\r\nX-Algorand-Location: http://127.0.0.2\r\nX-Algorand-Noderandom: cGVhMnBlYQ==\r\nX-Algorand-Version: 2.1\r\nX-Algorand-Genesis: private-v1\r\n\r\n"
2022-10-17T10:42:12.955967Z TRACE tokio_util::codec::framed_impl: flushing framed transport
2022-10-17T10:42:12.955978Z TRACE tokio_util::codec::framed_impl: writing; remaining=299
2022-10-17T10:42:12.955999Z TRACE tokio_util::codec::framed_impl: framed transport flushed
2022-10-17T10:42:12.956015Z DEBUG node{name="0"}: pea2pea::protocols::handshake: successfully handshaken with 127.0.0.1:54980
{"file":"wsPeer.go","function":"github.com/algorand/go-algorand/network.(*wsPeer).init","level":"debug","line":350,"msg":"wsPeer init outgoing=true \"127.0.0.1:54978\"","name":"127.0.0.1:0","time":"2022-10-17T12:42:12.956030+02:00"}
2022-10-17T10:42:12.956063Z DEBUG node{name="0"}: pea2pea::node: tentatively accepted a connection from 127.0.0.1:54981
{"file":"wsNetwork.go","function":"github.com/algorand/go-algorand/network.(*WebsocketNetwork).addPeer","level":"debug","line":2253,"msg":"ready","name":"127.0.0.1:0","time":"2022-10-17T12:42:12.956071+02:00"}
{"event":"ConnectedOut","file":"wsNetwork.go","function":"github.com/algorand/go-algorand/network.(*WebsocketNetwork).tryConnect","level":"info","line":2094,"local":"http://127.0.0.1:54979","msg":"Made outgoing connection to peer 127.0.0.1:54978","name":"127.0.0.1:0","rem
ote":"127.0.0.1:54978","time":"2022-10-17T12:42:12.956083+02:00"}
2022-10-17T10:42:12.956096Z DEBUG node{name="0"}: pea2pea::node: tentatively accepted a connection from 127.0.0.1:54982
{"details":{"Address":"127.0.0.1","HostName":"","Incoming":false,"InstanceName":"synth_node","Endpoint":"127.0.0.1:54978"},"file":"telemetry.go","function":"github.com/algorand/go-algorand/logging.(*telemetryState).logTelemetry","instanceName":"nASdpsibljD4Fq7Z","level":"
info","line":261,"msg":"/Network/ConnectPeer","name":"127.0.0.1:0","session":"","time":"2022-10-17T12:42:12.956094+02:00","v":"3.9.4"}
2022-10-17T10:42:12.956119Z DEBUG node{name="0"}: pea2pea::node: tentatively accepted a connection from 127.0.0.1:54984
2022-10-17T10:42:12.956144Z DEBUG node{name="0"}: pea2pea::node: tentatively accepted a connection from 127.0.0.1:54983
2022-10-17T10:42:12.956165Z DEBUG node{name="0"}: pea2pea::node: tentatively accepted a connection from 127.0.0.1:54985
2022-10-17T10:42:12.956184Z DEBUG node{name="0"}: pea2pea::node: tentatively accepted a connection from 127.0.0.1:54986
2022-10-17T10:42:12.956204Z DEBUG node{name="0"}: pea2pea::node: tentatively accepted a connection from 127.0.0.1:54987
2022-10-17T10:42:12.956204Z DEBUG node{name="0"}: pea2pea::node: tentatively accepted a connection from 127.0.0.1:54987
2022-10-17T10:42:12.956228Z DEBUG node{name="0"}: pea2pea::node: tentatively accepted a connection from 127.0.0.1:54988
2022-10-17T10:42:12.956247Z DEBUG node{name="0"}: pea2pea::node: tentatively accepted a connection from 127.0.0.1:54989
2022-10-17T10:42:12.956266Z DEBUG node{name="0"}: pea2pea::node: tentatively accepted a connection from 127.0.0.1:54991
2022-10-17T10:42:12.956284Z DEBUG node{name="0"}: pea2pea::node: tentatively accepted a connection from 127.0.0.1:54990
2022-10-17T10:42:12.956303Z DEBUG node{name="0"}: pea2pea::node: tentatively accepted a connection from 127.0.0.1:54992
2022-10-17T10:42:12.956322Z DEBUG node{name="0"}: pea2pea::node: tentatively accepted a connection from 127.0.0.1:54993
2022-10-17T10:42:12.956341Z DEBUG node{name="0"}: pea2pea::node: tentatively accepted a connection from 127.0.0.1:54994
2022-10-17T10:42:12.956377Z DEBUG node{name="0"}: pea2pea::node: tentatively accepted a connection from 127.0.0.1:54995
2022-10-17T10:42:12.956397Z DEBUG node{name="0"}: pea2pea::node: tentatively accepted a connection from 127.0.0.1:54996
2022-10-17T10:42:12.956559Z TRACE node{name="0"}: pea2pea::protocols::reading: spawned a task for processing messages from 127.0.0.1:54980
2022-10-17T10:42:12.956580Z DEBUG node{name="0"}: pea2pea::protocols::handshake: shaking hands with 127.0.0.1:54981 as the Responder
2022-10-17T10:42:12.956603Z DEBUG node{name="0"}: pea2pea::protocols::handshake: shaking hands with 127.0.0.1:54982 as the Responder
2022-10-17T10:42:12.956625Z DEBUG node{name="0"}: pea2pea::protocols::handshake: shaking hands with 127.0.0.1:54984 as the Responder
2022-10-17T10:42:12.956644Z DEBUG node{name="0"}: pea2pea::protocols::handshake: shaking hands with 127.0.0.1:54983 as the Responder
2022-10-17T10:42:12.956662Z DEBUG node{name="0"}: pea2pea::protocols::handshake: shaking hands with 127.0.0.1:54985 as the Responder
2022-10-17T10:42:12.956680Z DEBUG node{name="0"}: pea2pea::protocols::handshake: shaking hands with 127.0.0.1:54986 as the Responder
2022-10-17T10:42:12.956698Z DEBUG node{name="0"}: pea2pea::protocols::handshake: shaking hands with 127.0.0.1:54987 as the Responder
2022-10-17T10:42:12.956716Z DEBUG node{name="0"}: pea2pea::protocols::handshake: shaking hands with 127.0.0.1:54988 as the Responder
2022-10-17T10:42:12.956734Z DEBUG node{name="0"}: pea2pea::protocols::handshake: shaking hands with 127.0.0.1:54989 as the Responder
2022-10-17T10:42:12.956752Z DEBUG node{name="0"}: pea2pea::protocols::handshake: shaking hands with 127.0.0.1:54991 as the Responder
2022-10-17T10:42:12.956770Z DEBUG node{name="0"}: pea2pea::protocols::handshake: shaking hands with 127.0.0.1:54990 as the Responder
2022-10-17T10:42:12.956788Z DEBUG node{name="0"}: pea2pea::protocols::handshake: shaking hands with 127.0.0.1:54992 as the Responder
2022-10-17T10:42:12.956806Z DEBUG node{name="0"}: pea2pea::protocols::handshake: shaking hands with 127.0.0.1:54993 as the Responder
2022-10-17T10:42:12.956824Z DEBUG node{name="0"}: pea2pea::protocols::handshake: shaking hands with 127.0.0.1:54994 as the Responder
2022-10-17T10:42:12.956842Z DEBUG node{name="0"}: pea2pea::protocols::handshake: shaking hands with 127.0.0.1:54995 as the Responder
2022-10-17T10:42:12.956859Z DEBUG node{name="0"}: pea2pea::protocols::handshake: shaking hands with 127.0.0.1:54996 as the Responder
2022-10-17T10:42:12.956889Z TRACE node{name="0"}: pea2pea::protocols::reading: spawned a task for reading messages from 127.0.0.1:54980
2022-10-17T10:42:12.956931Z TRACE node{name="0"}: pea2pea::protocols::writing: spawned a task for writing messages to 127.0.0.1:54980
2022-10-17T10:42:12.956998Z TRACE tokio_util::codec::framed_impl: attempting to decode a frame
2022-10-17T10:42:12.957008Z TRACE tokio_util::codec::framed_impl: frame decoded from buffer
2022-10-17T10:42:12.957015Z  INFO node{name="0"}: ziggurat_algorand::protocol::handshake: 127.0.0.1:54981: received handshake message: b"GET /v1/private-v1/block/1 HTTP/1.1\r\nHost: 127.0.0.1:54978\r\nUser-Agent: algod/3.9 (stable; commit=921e8f6f; 4) darwin(arm64)\r\nAcc
ept-Encoding: gzip\r\n\r\n"
2022-10-17T10:42:12.957049Z ERROR node{name="0"}: ziggurat_algorand::protocol::handshake: missing Sec-WebSocket-Key!
2022-10-17T10:42:12.957074Z ERROR node{name="0"}: pea2pea::protocols::handshake: handshake with 127.0.0.1:54981 failed: invalid data
2022-10-17T10:42:12.957095Z TRACE tokio_util::codec::framed_impl: attempting to decode a frame
2022-10-17T10:42:12.957102Z TRACE tokio_util::codec::framed_impl: frame decoded from buffer
2022-10-17T10:42:12.957108Z  INFO node{name="0"}: ziggurat_algorand::protocol::handshake: 127.0.0.1:54982: received handshake message: b"GET /v1/private-v1/block/4 HTTP/1.1\r\nHost: 127.0.0.1:54978\r\nUser-Agent: algod/3.9 (stable; commit=921e8f6f; 4) darwin(arm64)\r\nAcc
ept-Encoding: gzip\r\n\r\n"
2022-10-17T10:42:12.957138Z ERROR node{name="0"}: ziggurat_algorand::protocol::handshake: missing Sec-WebSocket-Key!
{"Context":"sync","file":"universalFetcher.go","function":"github.com/algorand/go-algorand/catchup.(*HTTPFetcher).getBlockBytes","level":"debug","line":228,"msg":"GET \"http://127.0.0.1:54978/v1/private-v1/block/1\" : Get \"http://127.0.0.1:54978/v1/private-v1/block/1\":
EOF","name":"127.0.0.1:0","time":"2022-10-17T12:42:12.957160+02:00"}
2022-10-17T10:42:12.957158Z ERROR node{name="0"}: pea2pea::protocols::handshake: handshake with 127.0.0.1:54982 failed: invalid data
{"Context":"sync","file":"service.go","function":"github.com/algorand/go-algorand/catchup.(*Service).fetchAndWrite","level":"debug","line":253,"msg":"fetchAndWrite(1): Could not fetch: Get \"http://127.0.0.1:54978/v1/private-v1/block/1\": EOF (attempt 1)","name":"127.0.0.
1:0","time":"2022-10-17T12:42:12.957175+02:00"}
```

Reply:
Indeed. The node sends these GET block messages over HTTP to us, and it's not even waiting for a handshake to succeed. I'm planning to construct and send the same GET block message to the node to see how it will reply (when we skip the handshake procedure first)
```
Core","name":"127.0.0.1:0","time":"2022-10-17T12:49:26.081933+02:00"}
{"Context":"sync","file":"universalFetcher.go","function":"github.com/algorand/go-algorand/catchup.(*HTTPFetcher).getBlockBytes","level":"debug","line":228,"msg":"GET \"http://127.0.0.1:55035/v1/private-v1/block/2\" : Get \"http://127.0.0.1:55035/v1/private-v1/block/2\":
EOF","name":"127.0.0.1:0","time":"2022-10-17T12:49:26.082638+02:00"}
{"Context":"sync","file":"service.go","function":"github.com/algorand/go-algorand/catchup.(*Service).fetchAndWrite","level":"debug","line":253,"msg":"fetchAndWrite(2): Could not fetch: Get \"http://127.0.0.1:55035/v1/private-v1/block/2\": EOF (attempt 1)","name":"127.0.0.
1:0","time":"2022-10-17T12:49:26.082656+02:00"}
{"Context":"sync","file":"universalFetcher.go","function":"github.com/algorand/go-algorand/catchup.(*HTTPFetcher).getBlockBytes","level":"debug","line":228,"msg":"GET \"http://127.0.0.1:55035/v1/private-v1/block/1\" : Get \"http://127.0.0.1:55035/v1/private-v1/block/1\":
EOF","name":"127.0.0.1:0","time":"2022-10-17T12:49:26.082660+02:00"}
```